Run tests directly with pytest

fowlslegs opened the pull request:
Changes/ new features include:

- Easy runs of 1+ selected functional/unit tests, callable by short names.
- Pass any options through to pytest when running any unit or functional
  test(s), or execute arbitrary pytest commands in the test environment.
- Run docstring tests.
- Create non-admin journalists as well.
- Better exception handling and error reporting.
- Ensures cleanup of test environment _always_ happens.
- Some work towards Python 3 compatibility for eventual transition.
- Adds sick SD ASCII art to `./manage.py run`.
- Fix `crypto_util` module doctests.
- Update pytest section syntax in `setup.cfg`.

Supersedes https://github.com/freedomofpress/securedrop/pull/1516 and resolves https://github.com/freedomofpress/securedrop/issues/1412.

fowlslegs commented:
Adding "pending additional work" label until the following issues are resolved:
* Functional tests are passing if called individually, but not when called all at once (e.g., `./manage.py functional -a` will fail during `test_submit_and_retrieve_file`, while `./manage.py functional submit_and_retrieve_file` will pass).
* `./manage.py {unit, functional} -a` is not producing a proper coverage report.
* Doctests are not being executed correctly.

fowlslegs commented:
Pending CI tests passing, this PR should be ready for proper review. All issues in my last comment have been addressed.

fowlslegs commented:
Rebased.

fowlslegs commented:
Rebased again.

fowlslegs commented:
Rebased again.

fowlslegs commented:
Rebased.

fowlslegs commented:
Rebased.

fowlslegs commented:
Lmk when you want to review this anyone and I'll quickly push a rebase for you.

fowlslegs commented:
Rebased!

fowlslegs commented:
This PR is blocking https://github.com/freedomofpress/securedrop/issues/1507 (see https://github.com/freedomofpress/securedrop/pull/1612#issuecomment-285823600). Review would be much appreciated.

redshiftzero commented:
If we are going to refactor manage.py and remove `test.sh` - which is great - it seems like the cleanest approach would be to use `pytest` to run the tests. This would obviate the need for maintaining any special wrapper code in SecureDrop to run single tests, files of tests, etc. since these are actions one does directly in `pytest`. To do this, we would need to move the startup/teardown test code (e.g. `_start_test_rqworker` ) out of `manage.py` to `conftest.py` (perhaps using `pytest_configure`). What do you think?

fowlslegs commented:
@redshiftzero That sounds like a good idea. I've wanted to move the testing logic out of the `manage` module anyway. As annoying as it is not having something better than the current `manage` module in the short term, I'll get right to work on making those changes here. I'm going to ignore #1610 for now--that should be fixed by another PR.

conorsch commented:
Agreed with @redshiftzero on leveraging `conftest.py`. #1616 does the same for the config tests. (it _also_ provides a bash wrapper on top of that but that's necessary given the current implementation of the Ansible config.)

Perhaps we should merge this and tackle the `conftest.py` addition in a subsequent PR?

fowlslegs commented:
So I moved the testing directory out of the app code directory and made it a sibling directory of the app code directory. This is described in the pytest best practices documentation, and makes sense in the context of https://github.com/freedomofpress/securedrop/pull/1612#issuecomment-285823600, and also with regards to our plan to remove any notion of testing details from the app code, and to instead use dependency injection to create the appropriate environment for testing.

I also moved the `setUp` and `tearDown`-type functionality from `manage.py` into a `conftest.py`, which has been working fine for the unit tests when calling `pytest` in `tests/`. OTOH, I've been running into problems with the functional tests. I should note that even in `develop`, if you run the tests out of the order they are in in the `tests.sh` wrapper script, they will fail. In my original rewrite of `manage` I made it so they ran in that same order when one called `./manage.py functional`, deferring a proper fix. Now, regardless of order, 3/5 tests will fail.

The problem is _somewhere_ forking is occurring and this is causing `pycrypto` to throw an exception. The reason for this exception is that when a process is forked the parent and child process will have the same seed, so CSPRNG output will be identical. For w/e reason `pycrypto` will check for this problem, but instead of remedying it itself, it throws an error. I've yet to be able to track down (i) why this just started happening, (ii) where the fork is actually happening, and (iii) where to call `Crypto.Random.at_fork()` to resolve this issue (or if the issue is a deeper one and calling that will be a hacky fix). See error below, it's the same one that's causing all 3 of the failing tests:

<details>
  <summary>pytest error log</summary>
__________________________ SubmitAndRetrieveMessage.test_submit_and_retrieve_happy_path ___________________________

self = <tests.functional.test_submit_and_retrieve_message.SubmitAndRetrieveMessage testMethod=test_submit_and_retrieve_happy_path>

    def test_submit_and_retrieve_happy_path(self):
        self._source_visits_source_homepage()
>       self._source_chooses_to_submit_documents()

functional/test_submit_and_retrieve_message.py:22:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
functional/source_navigation_steps.py:42: in _source_chooses_to_submit_documents
    codename = self.driver.find_element_by_css_selector('#codename')
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py:437: in find_element_by_css_selector
    return self.find_element(by=By.CSS_SELECTOR, value=css_selector)
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py:752: in find_element
    'value': value})['value']
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py:236: in execute
    self.error_handler.check_response(response)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <selenium.webdriver.remote.errorhandler.ErrorHandler object at 0x7f28c5593590>
response = {'status': 500, 'value': '{"name":"findElement","sessionId":"7890dad4-e890-4a35-b2db-e06b4e26abb5","status":7,"value":...mp/tmpaDZDrJ/extensions/fxdriver@googlecode.com/components/driver-component.js","lineNumber":625,"columnNumber":5}]}}'}

    def check_response(self, response):
        """
            Checks that a JSON response from the WebDriver does not have an error.

            :Args:
             - response - The JSON response from the WebDriver server as a dictionary
               object.

            :Raises: If the response contains an error message.
            """
        status = response.get('status', None)
        if status is None or status == ErrorCode.SUCCESS:
            return

        value = None
        message = response.get("message", "")
        screen = response.get("screen", "")
        stacktrace = None
        if isinstance(status, int):
            value_json = response.get('value', None)
            if value_json and isinstance(value_json, basestring):
                import json
                try:
                    value = json.loads(value_json)
                    status = value.get('error', None)
                    if status is None:
                        status = value["status"]
                        message = value["value"]
                        if not isinstance(message, basestring):
                            value = message
                            try:
                                message = message['message']
                            except TypeError:
                                message = None
                    else:
                        message = value.get('message', None)
                except ValueError:
                    pass

        exception_class = ErrorInResponseException
        if status in ErrorCode.NO_SUCH_ELEMENT:
            exception_class = NoSuchElementException
        elif status in ErrorCode.NO_SUCH_FRAME:
            exception_class = NoSuchFrameException
        elif status in ErrorCode.NO_SUCH_WINDOW:
            exception_class = NoSuchWindowException
        elif status in ErrorCode.STALE_ELEMENT_REFERENCE:
            exception_class = StaleElementReferenceException
        elif status in ErrorCode.ELEMENT_NOT_VISIBLE:
            exception_class = ElementNotVisibleException
        elif status in ErrorCode.INVALID_ELEMENT_STATE:
            exception_class = InvalidElementStateException
        elif status in ErrorCode.INVALID_SELECTOR \
                or status in ErrorCode.INVALID_XPATH_SELECTOR \
                or status in ErrorCode.INVALID_XPATH_SELECTOR_RETURN_TYPER:
            exception_class = InvalidSelectorException
        elif status in ErrorCode.ELEMENT_IS_NOT_SELECTABLE:
            exception_class = ElementNotSelectableException
        elif status in ErrorCode.INVALID_COOKIE_DOMAIN:
            exception_class = WebDriverException
        elif status in ErrorCode.UNABLE_TO_SET_COOKIE:
            exception_class = WebDriverException
        elif status in ErrorCode.TIMEOUT:
            exception_class = TimeoutException
        elif status in ErrorCode.SCRIPT_TIMEOUT:
            exception_class = TimeoutException
        elif status in ErrorCode.UNKNOWN_ERROR:
            exception_class = WebDriverException
        elif status in ErrorCode.UNEXPECTED_ALERT_OPEN:
            exception_class = UnexpectedAlertPresentException
        elif status in ErrorCode.NO_ALERT_OPEN:
            exception_class = NoAlertPresentException
        elif status in ErrorCode.IME_NOT_AVAILABLE:
            exception_class = ImeNotAvailableException
        elif status in ErrorCode.IME_ENGINE_ACTIVATION_FAILED:
            exception_class = ImeActivationFailedException
        elif status in ErrorCode.MOVE_TARGET_OUT_OF_BOUNDS:
            exception_class = MoveTargetOutOfBoundsException
        else:
            exception_class = WebDriverException
        if value == '' or value is None:
            value = response['value']
        if isinstance(value, basestring):
            if exception_class == ErrorInResponseException:
                raise exception_class(response, value)
            raise exception_class(value)
        if message == "" and 'message' in value:
            message = value['message']

        screen = None
        if 'screen' in value:
            screen = value['screen']

        stacktrace = None
        if 'stackTrace' in value and value['stackTrace']:
            stacktrace = []
            try:
                for frame in value['stackTrace']:
                    line = self._value_or_default(frame, 'lineNumber', '')
                    file = self._value_or_default(frame, 'fileName', '<anonymous>')
                    if line:
                        file = "%s:%s" % (file, line)
                    meth = self._value_or_default(frame, 'methodName', '<anonymous>')
                    if 'className' in frame:
                        meth = "%s.%s" % (frame['className'], meth)
                    msg = "    at %s (%s)"
                    msg = msg % (meth, file)
                    stacktrace.append(msg)
            except TypeError:
                pass
        if exception_class == ErrorInResponseException:
            raise exception_class(response, message)
        elif exception_class == UnexpectedAlertPresentException and 'alert' in value:
            raise exception_class(message, screen, stacktrace, value['alert'].get('text'))
>       raise exception_class(message, screen, stacktrace)
E       NoSuchElementException: Message: Unable to locate element: {"method":"css selector","selector":"#codename"}
E       Stacktrace:
E           at FirefoxDriver.prototype.findElementInternal_ (file:///tmp/tmpaDZDrJ/extensions/fxdriver@googlecode.com/components/driver-component.js:10770)
E           at fxdriver.Timer.prototype.setTimeout/<.notify (file:///tmp/tmpaDZDrJ/extensions/fxdriver@googlecode.com/components/driver-component.js:625)

/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py:192: NoSuchElementException
---------------------------------------------- Captured stderr call -----------------------------------------------
 * Running on http://127.0.0.1:33346/ (Press CTRL+C to quit)
 * Running on http://127.0.0.1:46607/ (Press CTRL+C to quit)
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/css/source.css HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/gen/source.js?8c3b1397 HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/font-awesome/times-white.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/logo.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/font-awesome/cloud-upload-white.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/font-awesome/cloud-upload-blue.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/font-awesome/comments-white.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/font-awesome/comments-blue.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /static/i/toronion.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /generate HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 2000, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1991, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1567, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/vagrant/securedrop/source.py", line 163, in generate
    codename = generate_unique_codename()
  File "/vagrant/securedrop/source.py", line 134, in generate_unique_codename
    codename = crypto_util.genrandomid(num_words)
  File "/vagrant/securedrop/crypto_util.py", line 91, in genrandomid
    return ' '.join(random.choice(words) for x in range(words_in_random_id))
  File "/vagrant/securedrop/crypto_util.py", line 91, in <genexpr>
    return ' '.join(random.choice(words) for x in range(words_in_random_id))
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/random.py", line 102, in choice
    return seq[self.randrange(len(seq))]
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/random.py", line 83, in randrange
    r = self.getrandbits(size(num_choices))
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/random.py", line 51, in getrandbits
    return mask & bytes_to_long(self._randfunc(ceil_div(k, 8)))
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 202, in read
    return self._singleton.read(bytes)
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 178, in read
    return _UserFriendlyRNG.read(self, bytes)
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 137, in read
    self._check_pid()
  File "/usr/local/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 153, in _check_pid
    raise AssertionError("PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()")
AssertionError: PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /generate?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /generate?__debugger__=yes&cmd=resource&f=jquery.js HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:28] "GET /generate?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:29] "GET /generate?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:29] "GET /generate?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 -
127.0.0.1 - - [15/Mar/2017 20:42:29] "GET /generate?__debugger__=yes&cmd=resource&f=ubuntu.ttf HTTP/1.1" 200 -
======================================= 3 failed, 2 passed in 41.65 seconds =======================================
</details>

fowlslegs commented:
Issues resolved and solution presented in the last commit message:

>* Moves tests outside `securedrop/` dir to sibling dir as described in pytest
  best practices, as a step in the direction of our goal of completely
  separating test and application logic, and to ease implementation of
  #1612 (comment).
>* Tests are now called by invoking `pytest` directly instead of using
  `securedrop/manage.py`. The pytest-session-level setup and tear down
  functionality now lives in `tests/conftest.py`.
>* Fixes #1610. By adding `db_session.remove` to `utils.env.tearDown`, which is
  called by all `TestCase.tearDown` methods in our test suite, session scope is
  tied to test method scope. The next time a `Session` object is needed during
  the execution of a proceeding test method, the `scoped_session` registry
  (`db_session`) will call the `sessionmaker` factory as needed, which will
  create a new `Session` object and a new connection to the freshly re-created
  test database. This new `Session` object will be correctly synced with the
  database state, and test methods will execute with this clean context.
>* Calls `Crypto.Random.atfork()` in the `start_journalist_server` and
  `start_source_server` functional test `setUp` functions to re-seed the
  `pycrypto` CSPRNG in the Werkzeug server app child processes. This prevents
  the CSPRNG object from throwing an exception because these child processes
  have the same CSPRNG state as their parent, and will dangerously produce the
  same output.

fowlslegs commented:
And rebased.

fowlslegs commented:
I can't figure out why `pytest --doctest-glob='../securedrop/*.py'` won't run the `crypto_util` doctests. This isn't essential to merging, but it would be nice to add `--doctest-glob=../securedrop/*.py` to the `addopts=` line of `tests/pytest.ini` (i.e., to automatically run all doctests on invocation of `pytest` including in CI).

conorsch commented:
Looking good! Tested in the development VM by running:

```
cd /vagrant/tests
pytest -v
```

All tests are passing, and I like the output. We'll need to add docs for running the tests, but we can punt that to #1208, which will need lots of updates prior to releasing 0.4. @fowlslegs can you rebase one more time to resolve the `.travis.yml` conflict caused by merge of #1628?

redshiftzero commented on tests/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from os.path import abspath, dirname, join, realpath
+import sys
+
+sys.path.append(abspath(join(dirname(realpath(__file__)), '..', 'securedrop')))
```
Please add a comment to the code explaining why this is necessary - note I understand this is here because you moved the tests directory and you don't want to modify the imports in the tests, but it might not be straightforward to someone looking at the code for the first time

redshiftzero commented on tests/functional/functional_test.py:
```diff
@@ -65,13 +67,15 @@ def setUp(self):
         self.journalist_location = "http://localhost:%d" % journalist_port
 
         def start_source_server():
+            Random.atfork()
```
Can you add a short comment with maybe a link to some external documentation why this Random.atfork() was necessary?

conorsch commented:
Thanks for careful review, @redshiftzero! And @fowlslegs, great job hammering through this one—merging!
